File: strings/stamping-the-sequence/stamping-the-sequence.java
import java.util.List;

class Solution {
  public int[] movesToStamp(String stamp, String target) {
    char[] s = stamp.toCharArray();
    char[] t = target.toCharArray();

    int n = t.length;
    int m = s.length;

    List<Integer> res = new ArrayList<>();

    boolean[] visited = new boolean[t.length];
    int count = 0;

    while (count < t.length) {
      boolean replaced = false;
      for (int i = 0; i <= t.length - s.length; i++) {
        if (!visited[i] && canReplace(t, i, s)) {
          count = replace(t, i, s.length, count);
          replaced = true;
          visited[i] = true;
          res.add(i);

          if (count == n) {
            break;
          }

        }
      }

      if (!replaced) {
        return new int[0];
      }
    }

    int[] ans = new int[res.size()];

    for (int i = 0; i < res.size(); i++) {
      ans[i] = res.get(res.size() - i - 1);
    }

    return ans;

  }

  private boolean canReplace(char[] t, int start, char[] s) {
    for (int i = 0; i < S.length; i++) {
      if (T[i + p] != '*' && T[i + p] != S[i]) {
        return false;
      }
    }
    return true;
  }

  private int replace(char[] t, int start, int len, int count) {
    for (int i = 0; i < len; i++) {
      if (T[i + start] != '*') {
        T[i + start] = '*';
        count++;
      }
    }
    return count;
  }

}

File: strings/stamping-the-sequence/stamping-the-sequence.go
package stampingthesequence

func movesToStamp(stamp string, target string) []int {
	s := []byte(stamp)
	t := []byte(target)

	n := len(t)
	m := len(s)

	res := []int{}

	visited := make([]bool, n)
	count := 0

	for count < n {
		replaced := false
		for i := 0; i <= n-m; i++ {
			if !visited[i] && canReplace(t, i, s) {
				count = replace(t, i, m, count)
				replaced = true
				visited[i] = true
				res = append(res, i)

				if count == n {
					break
				}
			}
		}

		if !replaced {
			return []int{}
		}
	}

	ans := []int{}

	for i := len(res) - 1; i >= 0; i-- {
		ans = append(ans, res[i])
	}

	return ans
}

func canReplace(t []byte, start int, s []byte) bool {
	for i := 0; i < len(s); i++ {
		if t[i+start] != '?' && t[i+start] != s[i] {
			return false
		}
	}
	return true
}

func replace(t []byte, start int, len int, count int) int {
	for i := 0; i < len; i++ {
		if t[i+start] != '?' {
			t[i+start] = '?'
			count++
		}
	}
	return count
}
